cedar-policy-core/src/parser/cst_to_ast.rs

```diff
@@ -82,62 +85,67 @@ impl Node<Option<cst::Policies>> {
     /// corresponding generated `PolicyID`s
     pub fn with_generated_policyids(
         &self,
-    ) -> Option<impl Iterator<Item = (ast::PolicyID, &Node<Option<cst::Policy>>)>> {
+    ) -> Result<impl Iterator<Item = (ast::PolicyID, &Node<Option<cst::Policy>>)>, ParseErrors>
+    {
         // if `self` doesn't have data, nothing we can do here, just propagate
         // the `None`; we don't need to signal an error, because one was already
         // signaled when the `Node` without data was created
-        let policies = self.as_inner()?;
+        let policies = self.ok_or_missing()?;
```

Comment:
Comment is awkward now, it says "we don't need to signal an error", but it does signal an error.  Moreover the old code never threw `MissingNodeData` and the new code does; I'm not sure that's a desirable behavior change?

Reply:
> old code never threw MissingNodeData and the new code does

The `simple-parser` fuzz target asserts this fact, so we will at the very least need to update that. 


IIUC, the node data is `None` only when we use lalrpops error recovery mechanism. 
Which is only here
https://github.com/cedar-policy/cedar/blob/11ae14a8f1437b06d08e59ffcd4b535b8ca75f72/cedar-policy-core/src/parser/grammar.lalrpop#L192
and here
https://github.com/cedar-policy/cedar/blob/11ae14a8f1437b06d08e59ffcd4b535b8ca75f72/cedar-policy-core/src/parser/grammar.lalrpop#L115

If we wanted to we could probably refactor so that the CST nodes that don't use this mechanism can't be `None`.

Reply:
Ah, I meant to remove this comment throughout the code as I update to using `ok_or_missing`, but looks like I missed this one.

I think that throwing the error is the more sensible behavior, since the alternative could result in returning a `None` result, but no parse errors. Per my understanding, the `MissingNodeData` error can only be thrown if CST parsing failed, which should mean that the CST->AST conversion is never reached -- I'll update the error message for this variant from "data should not be empty" to "invalid input data: unable to continue processing" (open to better suggestions!)

Reply:
I'd also try to rename the error variant and `ok_or_missing` function to something more self explanatory. It's not at all clear why the data would be "Missing" just based on the name.

Reply:
> the node data is `None` only when we use lalrpops error recovery mechanism.

My knowledge is from a few refactors ago, but I didn't think this was the case.  The CST->AST transform frequently (last I looked) uses a pattern where it encounters a problem, and instead of immediately throwing the error, just adds the error to a `Vec`, makes the node `None`, and keeps going, in order to try to find other possible (unrelated) errors.  The old `.as_inner()` line here basically says "if we have a `None`, this method doesn't have any meaningful way to find more errors for you" and returns.

Reply:
I don't think that's right. It frequently returns `None`, but I don't think it every makes a CST node `None` that wasn't already

Reply:
Ack. You're right.

Reply:
I agree with John's comment: I don't think there is any way the text->CST conversion will return an empty node without an error. So I've changed the name & description of `MissingNodeData` to the following:
```rust
/// Returned when a CST node is empty during CST to AST/EST conversion.
/// This should have resulted in an error during the text to CST
/// conversion, which will terminate parsing. So it should be unreachable
/// in later stages.
#[error("internal invariant violated. Parsed data node should not be empty")]
EmptyNodeInvariantViolation,
```